hash/hash_comparator.go

```diff
@@ -260,3 +293,37 @@
 
 	return p, salt, hash, nil
 }
+
+var (
+	md5WithoutSalt = regexp.MustCompile(`^\$md5\$v=1\$([a-zA-Z0-9]{32})$`)
+	md5WithSalt    = regexp.MustCompile(`^\$md5\$v=1\$pf=(.*)\$([a-zA-Z0-9=]{1,})\$([a-zA-Z0-9]{32})$`)
+)
+
+// decodeMD5Hash decodes MD5 encoded password hash.
+// format without salt: $md5$<hash>
+// format with salt $md5$v=1$pf=<salting-format>$<salt-base64>$<hash>
+func decodeMD5Hash(encodedHash string) (pf, salt, hash []byte, err error) {
+	if parts := md5WithSalt.FindStringSubmatch(encodedHash); len(parts) == 4 {
+		salt, err := base64.StdEncoding.DecodeString(parts[2])
+		if err != nil {
+			return nil, nil, nil, errors.WithStack(err)
+		}
+
+		hash, err := hex.DecodeString(parts[3])
```

Comment:
I think we are using base64 encoding everywhere else in password format strings. Why use hex encoding here?

Reply:
Honestly, I'm in between on this subject. In the one hand almost all of the products I've seen to this day keeps MD5 hashes in hexadecimal string format. On the other hand using hex encoding here contradicts with out formats as you say. My main point of concern in here is using base64 encoding might be a little confusing for some users. If we ask them to give it in base64 encoding instead of hex string, following questions might be arose:

- Should the user give base64 output of hex string itself ? 
- Should the user give base64 output of bytes itself ? 

Currently on other formats, we are using the strategy of encoding bytes with base64, so i'm assuming in that scenario we will go with the second option. In that case almost all of the users should convert hexadecimal formatted string to bytes first, and then they should encode those bytes with base64

Reply:
Yes, I'd prefer that. The output of the MD5 hash itself should be base64 encoded.

Reply:
@hperl Alright then, just to be clear on this, users should be encoding 16 byte md5 hash with base64, right ? Not the 32 byte md5 hex string representation.

Reply:
yes :)